Common/cpp/NativeModules/NativeReanimatedModule.cpp

```diff
@@ -173,87 +160,62 @@ NativeReanimatedModule::NativeReanimatedModule(
-  auto viewDescriptorsSV = ShareableValue::adapt(rt, viewDescriptors, this);
-
+    const jsi::Value &worklet) {
+  auto shareableWorklet = extractShareableOrThrow(rt, worklet);
+  assert(
+      shareableWorklet->valueType == Shareable::WorkletType &&
+      "only worklets can be scheduled to run on UI");
+  auto uiRuntime = runtimeHelper->uiRuntime;
   scheduler->scheduleOnUI([=] {
-    auto mapperFunction =
-        mapperShareable->getValue(*runtime).asObject(*runtime).asFunction(
-            *runtime);
-    std::shared_ptr<jsi::Function> mapperFunctionPointer =
-        std::make_shared<jsi::Function>(std::move(mapperFunction));
-
-    std::shared_ptr<Mapper> mapperPointer = std::make_shared<Mapper>(
-        this,
-        newMapperId,
-        mapperFunctionPointer,
-        inputMutables,
-        outputMutables);
-    if (optimalizationLvl > 0) {
-      mapperPointer->enableFastMode(
-          optimalizationLvl, updaterSV, viewDescriptorsSV);
-    }
-    mapperRegistry->startMapper(mapperPointer);
-    maybeRequestRender();
+    jsi::Runtime &rt = *uiRuntime;
+    auto workletValue = shareableWorklet->getJSValue(rt);
+    workletValue.asObject(rt).asFunction(rt).call(rt);
   });
-
-  return jsi::Value(static_cast<double>(newMapperId));
 }
 
-void NativeReanimatedModule::stopMapper(
+jsi::Value NativeReanimatedModule::makeShareableClone(
     jsi::Runtime &rt,
-    const jsi::Value &mapperId) {
-  unsigned long id = mapperId.asNumber();
-  scheduler->scheduleOnUI([=] {
-    mapperRegistry->stopMapper(id);
-    maybeRequestRender();
-  });
+    const jsi::Value &value) {
+  std::shared_ptr<Shareable> shareable;
+  if (value.isObject()) {
+    auto object = value.asObject(rt);
+    if (!object.getProperty(rt, "__workletHash").isUndefined()) {
+      shareable =
+          std::make_shared<ShareableWorklet>(runtimeHelper.get(), rt, object);
+    } else if (!object.getProperty(rt, "__init").isUndefined()) {
+      shareable =
+          std::make_shared<ShareableHandle>(runtimeHelper.get(), rt, object);
+    } else if (object.isFunction(rt)) {
+      shareable = std::make_shared<ShareableRemoteFunction>(
+          runtimeHelper.get(), rt, object.asFunction(rt));
+    } else if (object.isArray(rt)) {
+      shareable = std::make_shared<ShareableArray>(rt, object.asArray(rt));
+    } else {
+      shareable = std::make_shared<ShareableObject>(rt, object);
+    }
+  } else if (value.isString()) {
+    shareable = std::make_shared<ShareableString>(rt, value.asString(rt));
+  } else if (value.isUndefined()) {
+    shareable = std::make_shared<ShareableScalar>();
+  } else if (value.isNull()) {
+    shareable = std::make_shared<ShareableScalar>(nullptr);
+  } else if (value.isBool()) {
+    shareable = std::make_shared<ShareableScalar>(value.getBool());
+  } else if (value.isNumber()) {
+    shareable = std::make_shared<ShareableScalar>(value.getNumber());
+  } else {
+    throw "something wacky";
```

Comment:
e.g. `jsi::BigInt` or `jsi::ArrayBuffer`, maybe we could also share those between the JS and UI context?

Reply:
these types, I'm afraid, will fall into the branch that checks for "isObject". It makes sense, however, to make this error more descriptive anyways, even though I don't know under what circumstances it can be reached.